traits/trait_dict_object.py

```diff
@@ -98,232 +512,151 @@ class TraitDictObject(dict):
+    def value_validator(self, value):
+        """ Calls the trait's value_handler.validate
 
-        dict.__delitem__(self, key)
+        Parameters
+        ----------
+        value : any
+            The value to validate.
 
-        if self.name_items is not None:
-            self._send_trait_items_event(
-                self.name_items, TraitDictEvent(removed=removed)
-            )
+        Returns
+        -------
+        validated_value : any
+            The validated value.
 
-    def clear(self):
-        if len(self) > 0:
-            if self.name_items is not None:
-                removed = self.copy()
+        Raises
+        ------
+        TraitError
+            If the validation fails.
 
-            dict.clear(self)
-
-            if self.name_items is not None:
-                self._send_trait_items_event(
-                    self.name_items, TraitDictEvent(removed=removed)
-                )
+        """
+        trait = getattr(self, 'trait', None)
+        object_ref = getattr(self, 'object', None)
+        if object_ref is None or trait is None:
+            return value
 
-    def update(self, dic):
-        trait = getattr(self, "trait", None)
-        if trait is None:
-            dict.update(self, dic)
-            return
+        object = object_ref()
 
-        if len(dic) > 0:
-            new_dic = self._validate_dic(dic)
+        validate = trait.value_handler.validate
+        if validate is None:
+            return value
 
-            if self.name_items is not None:
-                added = {}
-                changed = {}
-                for key, value in new_dic.items():
-                    if key in self:
-                        changed[key] = self[key]
-                    else:
-                        added[key] = value
+        return validate(object, self.name, value)
 
-                dict.update(self, new_dic)
+    def notifier(self, trait_dict, added, changed, removed):
+        """ Fire the TraitDictEvent with the provided parameters.
 
-                self._send_trait_items_event(
-                    self.name_items,
-                    TraitDictEvent(added=added, changed=changed),
-                )
-            else:
-                dict.update(self, new_dic)
+        Parameters
+        ----------
+        trait_dict : dict
+            The complete dictionary.
+        added : dict
+            Dict of added items.
+        changed : dict
+            Dict of changed items.
+        removed : dict
+            Dict of removed items.
 
-    def setdefault(self, key, value=None):
-        if key in self:
-            return self[key]
+        Returns
+        -------
+        None
 
-        self[key] = value
-        result = self[key]
+        """
+        trait = getattr(self, 'trait', None)
+        name_items = getattr(self, 'name_items', None)
+        object_ref = getattr(self, 'object', None)
+        if trait is None or name_items is None or object_ref is None:
```

Comment:
Same here, could we have a test that fails if these lines were removed please?

Reply:
Checked that after renaming `key_validator` and `value_validator` (so that the code is actually run), removing this block of code still does not cause any tests to fail. I tested this on Python 3.8 and 3.6, investigating.

Reply:
See #990 which is targeting `TraitSetObject`, but it addresses more or less the same comment.

Reply:
Merged #990

Reply:
Resolved by #995